File: leetcode46.py
class Solution:
    def permute(self, nums:list):
        if len(nums) == 1:
            return [nums]
        all_list = []
        for a in nums:
            num_cpy = nums.copy()
            num_cpy.remove(a)
            nlist = self.permute(num_cpy)
            for l in nlist:
                l.append(a)
                all_list.append(l)
        return all_list

a = Solution()
print(a.permute([1,2,3]))

File: leetcode46.go
// +build ignore

package main

import "fmt"

func main() {
	fmt.Printf("%v", permute([]int{1, 2, 3}))
}
func backtrace(nums []int, res [][]int, i int) [][]int {
	if i == len(nums) {
		tmp2 := make([]int, len(nums))
		copy(tmp2, nums)
		tmp := append(res, tmp2)
		return tmp
	}
	for j := i; j < len(nums); j++ {
		temp := nums[i]
		nums[i] = nums[j]
		nums[j] = temp
		res = backtrace(nums, res, i+1)
		nums[j] = nums[i]
		nums[i] = temp
	}
	return res
}
func permute(nums []int) [][]int {
	result := make([][]int, 0)
	return backtrace(nums, result, 0)
}
